[WIP] Resolve overloads through the constraint system

harlanhaskins opened the pull request:
This patch tracks the rewrite of the overload resolution engine in terms of our new constraint solver.

Ideally, this overload resolver will punish coercions in order to arrive at the most optimal overload for a given function call.

harlanhaskins committed:
Start stubbing out overload resolution via the constraint solver.

harlanhaskins committed:
Reworked existing overload resolution code in terms of OverloadResolver.

harlanhaskins commented:
Aw yeah, the build is super broken right now! 😄

segiddins commented:
Lol I'm guessing you broke compiling the stdlib

harlanhaskins commented:
All overload resolution is broken because the constraint generator is relying on decls being wired up. Not sure how to fix this just yet...

harlanhaskins committed:
Clean up method call detection and resolution error handling

harlanhaskins committed:
Properly model severities and resolving possible ambiguities.

harlanhaskins committed:
Almost finished overload resolution algorithm

CodaFi commented on Sources/Base/ASTContext.swift:
```diff
@@ -224,83 +224,6 @@ public class ASTContext {
   }
 
-  /// - Returns: Whether the expression's type was changed
-  @discardableResult
```
Praise be, praise be

segiddins commented on Sources/Base/Types.swift:
```diff
@@ -10,38 +10,101 @@ enum FloatingPointType {
+  case integerLiteral
+
+  /// A data type representing a `nil` literal that hasn't been resolved to
+  /// a concrete pointer type.
```
`indirect` or `reference`

harlanhaskins replied:
Good call

segiddins commented on Sources/Base/Types.swift:
```diff
@@ -10,38 +10,101 @@ enum FloatingPointType {
   indirect case pointer(type: DataType)
+
+  /// An explicitly-sized array of a given type.
   indirect case array(field: DataType, length: Int?)
```
🔥

segiddins commented on Sources/Import/ClangImporter.swift:
```diff
@@ -477,7 +477,9 @@ class ClangImporter: Pass {
     case .number(let value, let raw):
-      return NumExpr(value: value, raw: raw, sourceRange: range)
+      let expr = NumExpr(value: value, raw: raw, sourceRange: range)
+      expr.type = .int64
```
will the handling for the suffixes overwrite this?

harlanhaskins replied:
Nope, it won't. Hmm... I'm going to rewrite this in terms of `Cursor.evaluate()` sometime soon

segiddins commented on Sources/Semantic Analysis/ConstraintSolution.swift:
```diff
@@ -14,6 +14,9 @@ enum CoercionKind: Int {
   /// An explicit type variable has been reified to a concrete type.
   case genericPromotion
 
+  /// A stirng literal has been promoted to *Int8.
```
`string`

harlanhaskins replied:
😂

segiddins commented on Sources/Semantic Analysis/ConstraintSolver.swift:
```diff
@@ -110,9 +119,46 @@ struct ConstraintSolver {
+        do {
+          return try solveSystem(system)
+        } catch {
+          break
```
😭 this makes breakpoints on throws harder

segiddins commented on Sources/Semantic Analysis/ConstraintSystem.swift:
```diff
@@ -77,7 +110,9 @@ struct ConstraintSystem {
         print("\(t1) == \(t2)")
       case let .conforms(t1, t2):
-        print("\(t1): \(t2)")
+        print("\(t1) :  \(t2)")
```
why the extra space?

harlanhaskins replied:
¯\_(ツ)_/¯

segiddins commented on Sources/Semantic Analysis/OverloadResolver.swift:
```diff
@@ -114,12 +150,26 @@ struct OverloadResolver {
+    return resolve(expr.args, candidates: candidates) { candidate in
+      expr.decl = candidate
+      csGen.visitSubscriptExpr(expr)
+      expr.decl = nil
```
put this in a `defer`

segiddins commented on Sources/Semantic Analysis/OverloadResolver.swift:
```diff
@@ -231,6 +278,11 @@ struct OverloadResolver {
     }
 
+    for solution in minSolutionCandidates {
+      print("solution for \(solution.chosenDecl.formattedName)")
```
debugging code

segiddins commented on Sources/Semantic Analysis/Sema.swift:
```diff
@@ -134,30 +153,30 @@ class Sema: ASTTransformer, Pass {
     let oldVarBindings = varBindings
+    let oldEnv = env
     super.withScope(e, f)
+    env = oldEnv
```
`defer`

segiddins commented on Sources/Semantic Analysis/Sema.swift:
```diff
@@ -168,26 +187,21 @@ class Sema: ASTTransformer, Pass {
+      let canRhs = context.canonicalType(rhs.type)
       if case .void = canRhs {
-        error(SemaError.incompleteTypeAccess(type: type, operation: "assign value from"),
+        error(SemaError.incompleteTypeAccess(type: canRhs,
```
shouldn't this be the sugared type?

harlanhaskins committed:
Overload resolution now works!

harlanhaskins committed:
Make no-context-nil explicitly ambiguous and address comments.

harlanhaskins committed:
Represent `Any` as an empty protocol composition

This work will be expanded in the future to allow for explicitly spelled-out protocol compositions.

harlanhaskins committed:
Properly mangle/demangle and generate constraints for protocol compositions.

harlanhaskins committed:
Properly handle existential downcasts.

harlanhaskins committed:
Trying to fix IRGen for overloads

harlanhaskins committed:
Fix propagation for return statements and actually solve ternary express

harlanhaskins committed:
Fix ambiguities for equality operators

harlanhaskins committed:
Fall back literal types in array and tuple expressions

harlanhaskins committed:
Solve tuple elements in the ConstraintSolver. Oops.

harlanhaskins committed:
Small cleanup

harlanhaskins committed:
Rename `coercion` constraint to `conversion` as that more explicitly represents the constraint.

segiddins commented:
To summarize the state of this PR:
Harlan and I have discussed how to fix the regressions, and the solution we've converged upon is that type solving needs to be done at the _statement_, rather than _expression_, level, to allow for cases in which expressions have sub-expressions with multiple possible types -- (see for example what happens with ternaries). For this to happen, the overload resolution code has to be juggled about to play nicely with expressions not being fully types until the enclosing statement has been fully visited.

CodaFi commented:
> type solving needs to be done at the statement, rather than expression, level, to allow for cases in which expressions have sub-expressions with multiple possible types -- (see for example what happens with ternaries)

The rule for this is that you type check both branches and if they yield two disjunctions you can either force the user to specify a contextual type or have some merge heuristic (try to take an intersection and see if you get a singular set, maybe?).  Otherwise you can use whatever branch that doesn't have a disjunct in it as a contextual type and keep solving.

Solving at the statement level should only be done if you intend to start propagating solution sets *between statements*.
